Add filtering for projects, based on hardward dependencies

abridger opened the pull request:
* Adds dependencies for specific projects in `src/assets/projects/projects.json`
* Adds filtering method in `kw-view-projects` to check whether the user is on a kit, and filter out projects that require unsupported hardware or features

Trello card: https://trello.com/c/KOuLMP9o/1075-filter-mic-camera-and-tts-challenges-from-projects-on-the-kit

Commit:
Add filtering for projects, based on hardward dependencies

* Adds dependencies for specific projects in `src/assets/projects/projects.json`
* Adds filtering method in `kw-view-projects` to check whether the user is on a kit, and filter out projects that require unsupported hardware or features

Trello card: https://trello.com/c/KOuLMP9o/1075-filter-mic-camera-and-tts-challenges-from-projects-on-the-kit

Commit:
Merge branch 'master' of github.com:KanoComputing/world into project-filtering

paulvarache commented on src/elements/kw-view-projects/kw-view-projects.html:
```diff
@@ -251,12 +251,50 @@
+                        if (project.dependencies) {
+                            project.dependencies.forEach((dependency) => {
+                                if (availableDependencies.indexOf(dependency) === -1) {
+                                    allDependenciesAvailable = false;
```
```js
allDependenciesAvailable = (availableDependencies.indexOf(dependency) !== -1)
```
?

abridger replied:
This should be `=== -1` since it's checking for unsupported dependencies and invalidating `allDependenciesAvailable` if any of the required dependencies are unsupported.

paulvarache reviewed (commented):
I know that I had some problems with low end devices on this categories rendering. I didn't run a timeline on this solution, but maybe using the built-in `filter` attribute for `dom-repeat` would be more efficient since it runs the filtering at the optimal moment for rendering

Commit:
Update filtering method to use the built in `filter` attribute for `dom-repeat`

Also adds the available dependencies to the view to avoid having to check the `userAgent` and returning the `availableDependencies` on every iteration of the category filter.

abridger commented:
Thanks for the tip! I've updated the view to use the `filter` attribute on the `dom-repeat`.

Commit:
Add text-to-speech dependency to Hello World challenge

paulvarache commented on src/elements/kw-view-projects/kw-view-projects.html:
```diff
@@ -251,6 +255,36 @@
+                    let allDependenciesAvailable = true;
+                    if (project.dependencies) {
+                        project.dependencies.forEach((dependency) => {
+                            if (this.availableDependencies.indexOf(dependency) === -1) {
```
Here using an `if` to set a variable to `false` could be simplified by setting the variable to the test condition of the `if`
```js
allDependenciesAvailable = this.availableDependencies.indexOf(dependency) !== -1
```

paulvarache replied:
Now that I think of it, for this kind of tasks,`every` on the array does exactly that.

paulvarache replied:
Fast reference to MDN if you need it: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every

Commit:
Move Dependency checking and filtering behavior to separate ProjectListingBehavior

* Allows the filtering and dependency checks to be used by the `kw-projects-header`
* Adds missing microphone depedency to project in `kw-projects-header`

paulvarache commented on src/elements/kano-projects-listing/kano-projects-listing-behavior.html:
```diff
@@ -0,0 +1,66 @@
+<script>
+    (function (Kano) {
+
+        window.dataLayer = window.dataLayer || [];
```
Small issue there :)

abridger replied:
Bother, sorry about that!

paulvarache reviewed (commented):
Looks pretty good

Commit:
Remove erroneous dataLayer reference
